### ISSUE TITLE
State of JS 2022 Preliminary Discussions

### ISSUE THREAD

**Issue by SachaG** [Devographics/surveys#47]
How can we improve the 2022 edition of the survey?

See also previous discussion here: https://github.com/Devographics/Monorepo/issues/61

### Things to Consider

- Sections to remove/add?
- Features questions to remove/add?
- Libraries questions to remove/add?
- Other questions to remove/add?
- Options to remove/add inside specific questions?
- Question types to change?
- Other improvements?

### 2021 Edition for Reference

- https://survey.devographics.com/survey/state-of-js/2021

### General Resources

- https://frontendmastery.com/posts/the-new-wave-of-javascript-web-frameworks/

**Comment by SachaG** [comment]
Some new items to consider, in no particular order: 

- https://zod.dev/
- https://slinkity.dev/
- LinkedIn Learning
- Stencil
- Qwik
- https://parceljs.org/blog/parcel-css/
- htmx
- https://markojs.com/
- https://shoelace.style/
- Selenium (https://www.npmjs.com/package/selenium-webdriver)

New questions to consider:

- Question about SSR/CSR/SSG etc. patterns
- Question about what people use JS for (https://github.com/Devographics/Monorepo/issues/61#issuecomment-1047961596)
-

**Comment by zackarychapple** [comment]
It would be really awesome to see some questions around Module Federation, in particular how much folks are using it and with what frameworks / stacks they are using it.  @ScriptedAlchemy may have som ideas on good questions on this.

**Comment by SachaG** [comment]
@zackarychapple it might be a bit too niche? At least I personally hadn't heard about it before now.

**Comment by RATIU5** [comment]
[Roadmap.sh](https://roadmap.sh/) has some pretty good learning resources/structure for JavaScript. 
I’m looking most forward to seeing how all these new frameworks are growing/declining/looking forward to.

### Routers
I want to see how developers have stances on routing. Which are better/most used/best dx.
- [Flamethrower](https://github.com/fireship-io/flamethrower)
- [Qwik City](https://github.com/BuilderIO/qwik/tree/main/packages/qwik-city)
- [Next.js](https://github.com/vercel/next.js)
- Others of course

### Data loading
It would be great to see a section on types of hydrating such as with islands, resumability, partial-hydration

Let notes, I personally don’t care if it’s a longer survey, I just love seeing all the new trends and direction JS is going.

**Comment by thetutlage** [comment]
I think I would love to see if having a dedicated section for backend frameworks and not mixing that with Nuxt, or Next or Svelte Kit.

And as @RomainLanz said, AdonisJS has been in the top list of "Other backend frameworks" for a couple of years, and yet missing in the selection option, because we are compared with Next, or Nuxt, even though the use cases these frameworks serve are different

**Comment by kettanaito** [comment]
Q: How do you mock requests in testing/development? (Multi-answer question)

Options:
- Not using any
- Nock
- MirageJS
- PollyJS
- Mock Service Worker
- cy.intercept
- Playwright's `route`
- Else (write in a freeform input)

I've been developing MSW (https://github.com/mswjs/msw) for almost five years now. It's one of the most commonly used and widely adopted API mocking solutions in JavaScript. It'd mean a lot to me to be able to see the analytics of usage on your survey. Consider including it.

**Comment by arendjr** [comment]
I realize this may be almost a personal crusade of mine, but the [pipeline operator proposal](https://github.com/tc39/proposal-pipeline-operator) uses the 2020 State of JS  survey as motivation for why adding the pipeline operator to the language is desired. However, back then the pipeline proposal seemed like it still had a chance of landing in the form of the F# proposal.

Nowadays, the pipeline operator looks like it may ascent using the Hack proposal instead. Unfortunately, many of the people who originally wanted a pipeline operator feel that it is better to **not** have a pipeline operator than the current Hack proposal. As such, we strongly feel the current proposal has lost sight of what we wanted and why we wanted it, yet uses an outdated State of JS survey as justification for why it is desired.

I know this might turn it into a rather niche question, but the most objective way we have of getting a true measure of how many people are still on board with the *current* proposal would be to present this question in the new JS survey. The precise phrasing of the question will still be difficult, as I notice when discussing the operator that people are often not aware the current proposal is the Hack proposal, which diverges significantly from how the operator behaves in other languages. But I hope the following phrasing would be objective enough without leading people to the answer:

Do you support the current [expression-based pipeline operator proposal](https://github.com/tc39/proposal-pipeline-operator#this-proposal-hack-pipes)?
* Yes
* No
* I don’t know

**Comment by larsthorup** [comment]
Testing tool option:
* https://vitest.dev/

**Comment by gabssnake** [comment]
I want to know if people are using SonarQube/SonarLint or they awareness. More broadly, maybe ask about code quality tools and security scanning tools.

**Comment by webbertakken** [comment]
Perhaps a little related to @arendjr's question
> Do you use WebAssembly with JavaScript, and if so, which language?

I am also very interested to hear what languages people interoperate with in JavaScript. Use cases could vary more widely than just WebAssembly though. Use cases could include:

- WebAssembly 
- Server-side or in-container processing of some heavier operations 
- On the client, to interact with embedded content like Unity player or WASM content
- Directly integrating with another language to compile the application, like with Elixir
- (more?)

Therefore I'd rephrase @arendjr's question to something slightly more generic:

"Interoperability - Do you ever invoke other languages from JavaScript? If yes, which ones (check all boxes that apply)"

- WebAssembly
- Rust
- Go
- Python
- C#
- C/C++
- Elixir
- Embedded content (e.g. Unity Player)
- Other _(add your option)_

**Comment by nandeshwarshubh** [comment]
It would be great to know how many people are using microfrontends and what underlying framework/plugin they are using for the same. Some of them being -
1. Webpack 5 and Module Federation - https://webpack.js.org/concepts/module-federation/
2. Single SPA - https://single-spa.js.org/
3. Qiankun - https://qiankun.umijs.org/
4. Luigi - https://luigi-project.io/
5. Frintjs - https://frint.js.org/
6. PuzzleJs - https://github.com/puzzle-js/puzzle-js
7. Podium - https://podium-lib.io/
8. Bit - https://bit.dev/

**Comment by csaltos** [comment]
How about adding Elm as a framework to consider in the new survey

**Comment by SnowSuno** [comment]
Would like to add in [trpc](https://trpc.io/) if possible, which rapidly gaining attention in github star count.

**Comment by SachaG** [comment]
> Do you support the current [expression-based pipeline operator proposal](https://github.com/tc39/proposal-pipeline-operator#this-proposal-hack-pipes)?
> 
> * Yes
> * No
> * I don’t know

I feel like this whole issue might be a little too hard to explain in the context of the survey and I'm not sure if people would be able to make an informed choice unless they're already very familiar with the matter? 

> Hey there! 👋🏻
> 
> I believe adding [AdonisJS](https://adonisjs.com/) is not a bad idea. We have been forgotten for years.

I'll definitely consider Adonis, I know it always comes high in the freeform submissions.

**Comment by anodynos** [comment]
I dare say I think there's a dichotomy that we got wrong: **Frontend VS Backend Frameworks**, especially when we're looking at the latter. 

Let me explain:

- **Frontend Frameworks** are easily defined: it's A) the UI, i.e what the user sees on screen AND B) it runs on the browser. 

- **Backend Frameworks** are harder to grasp, as they are currently defined as "running on the server", and there's our omission: 
  - we have **API-authoring** & **Backend services** frameworks (eg the bare bones Expressjs or the full fledged NestJS). These deal with accessing our DBs, Queues and other Cloud services, and most importantly offering an API gateway (eg REST, GraphQL, WebServices etc) to our app at run time. 
  - we also the static rendering frameworks like NextJS or Nuxt, who basically deal with static rendering of the UI, making your website blazing fast to load & SEO friendly. 
 
It's important to know, these 2 "backend" services are offering a completely different functionality on your tech stack and quite often you need both of them at some level, to create a production App. There's no choice really. Using something like [nextjs api-routes](https://nextjs.org/learn/basics/api-routes/creating-api-routes) is like using Visual Basic 1.0 to build a modern App. 

I could argue that the "Static Renderers" category is a very important one, and since it's closer to the UI it could use it's own "Frontend Delivery Frameworks" category under libraries. 

I could also argue that "Backend for API  & Services" is also a hugely important aspect, and could have it's own category. If we do both, the "Backend Frameworks" category kinda vanishes.

There's my 2c. Let's have this conversation and see where it leads :-)

**Comment by McSneaky** [comment]
Runtimes would be interesting, especially server ones. In browsers it's all kinda united, don't think there's point of listing Browser runtimes 🤔 

**Server:** 
- NodeJS
- Deno
- Bun
- ES4X
- JustJS

**IoT:** 
- JerryScript
- Elk
- IoT.js
- Jhonny-Five

**Edge:**
- AWS Lambda
- CloudFlare Workers
- Fastly Edge Compute
- Google Cloud Functions
- Akamai EdgeWorkers

**Comment by daniseguraf** [comment]
Hi there, It would be great to talk about used and satisfaction with  components libraries like:

-  [Material UI](https://mui.com/)
- [React Bootstrap](https://react-bootstrap.github.io/)
- [Ant Design](https://ant.design/components/overview/)
- [Mantine](https://mantine.dev/)
- [Tailwind UI](https://tailwindui.com/)

There are many more out there, what would you add to the list?

**Comment by SachaG** [comment]
@daniseguraf actually those are already covered in the [State of CSS](https://survey.devographics.com/survey/state-of-css/2022) survey.

**Comment by mischah** [comment]
Topic: MPAs (vs SPAs?)

Usage of new-ish SRR Patterns (Island Architecture, Partial Hydration, HTT Streaming etc) 
See: https://www.patterns.dev/posts/rendering-introduction/

And Usage of related libraries/frameworks: [Qwik](https://github.com/BuilderIO/qwik), [Astro](https://github.com/withastro/astro), [Elder](https://github.com/Elderjs/elderjs), [Marko](https://github.com/marko-js/marko),

**Comment by sirisian** [comment]
In the "Features missing from Javascript" it would be more meaningful to separate it into "Static Typing: Types as comments" and "Static Typing: Runtime/native types" (or something similar/concise). I've noticed the two groups not being on the same page at all. The former want type annotations for documentation that are stripped out before compiling and the latter want types as a language feature - with other associated type features and runtime type features.

**Comment by zackarychapple** [comment]
> @zackarychapple it might be a bit too niche? At least I personally hadn't heard about it before now.

Possibly I do know from what I've seen, the React, Next, and Angular communities are all frequently talking about it and it is something getting a lot of growth right now.

**Comment by jessealama** [comment]
Perhaps a quick pitch for [Decimal](https://www.proposals.es/proposals/Decimal)? There's a lot of interest out there in having arbitrary-precision, or at least high precision, decimal arithmetic. It's a language feature in Stage 2.

**Comment by arendjr** [comment]
> I feel like this whole issue might be a little too hard to explain in the context of the survey and I'm not sure if people would be able to make an informed choice unless they're already very familiar with the matter?

I think you’re right. A better approach might be to not put it in a separate question, but to simply split the answers for the Most Desired Features question: Which features do you feel JavaScript could use the most?

* Pipeline operator (Hack proposal)
* Pipeline operator (other proposal)
* … (other options)

That way it’s at least clear people are voting for a specific proposal, and people are less likely to vote for it under the assumption it would be similar to the pipeline operator in other languages.

**Comment by SachaG** [comment]
@anodynos I think this is a valid point, but then where would you put something like Redwood? https://redwoodjs.com/

**Comment by SachaG** [comment]
Thanks for your feedback everybody! Here's a preview of the survey outline: 

https://github.com/Devographics/surveys/issues/62

It's still a work in progress, so the things I've added/removed are not final by any means.

**Comment by SachaG** [comment]
I'll close this thread so we can discuss the survey preview directly in the other thread.

**Comment by anodynos** [comment]
@SachaG thanks for your reply. 

I'd say things like [*redwoodjs*](https://redwoodjs.com) and [*remix*](https://remix.run) cant always be described with one word. They are kind of meta, mini or maxi frameworks or stacks or mashups, often with some extra sauce. Remember all these acronyms we had like in 2010's with LAMP or MEAN etc. Often an evolution of those.

They are aiming to soften a particular angle, for example a "full stack" out of the box (like sailsjs) or generator/scafolding/cover setups for easy getting up and running, or make development easier & more effective in some aspect (eg UI) etc. 

They are mixing popular technologies and making essentially, new meta-frameworks. Or in the case of React (that is a more of a library, rather than an opinionated framework like Angular) they are making a framework out of it! 

Maybe we need a category like Meta Frameworks? Or are we going to be sued? LOL

**Comment by RATIU5** [comment]
In terms of "Meta Frameworks", I did not see QwikCity (The best DX I've used for FS) and SolidStart on the list of rendering frameworks. I understand they may not be on the list due to them not currently having a stable release version yet.

**Comment by SachaG** [comment]
@anodynos I agree the proper category would be "meta-framework" for something like Redwood, but then we would have a category with a single item… at least we'd know who will be number one! ;)

@RATIU5 yes they both feel a bit too new to be included this year. But who knows, we'll see.

**Comment by anodynos** [comment]
@SachaG I'm sure many *Meta frameworks* will surface if we have it as a category. 

But anyway I'm happy we've split *Backend Services* from *Backend Renderers*, cause they solve very different problems.
